api/pkg/apis/projectcalico/v3/blockaffinity.go

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2022 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v3
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	KindBlockAffinity     = "BlockAffinity"
+	KindBlockAffinityList = "BlockAffinityList"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// BlockAffinity maintains a block affinity's state
+type BlockAffinity struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// Specification of the BlockAffinity.
+	Spec BlockAffinitySpec `json:"spec,omitempty"`
+}
+
+// BlockAffinitySpec contains the specification for a BlockAffinity resource.
+type BlockAffinitySpec struct {
+	// The state of the block affinity with regard to any referenced IPAM blocks.
+	State string `json:"state"`
+
+	// The node that this block affinity is assigned to.
+	Node string `json:"node"`
+
+	// The CIDR range this block affinity references.
+	CIDR string `json:"cidr"`
+
+	// Deleted indicates that this block affinity is deleted.
```

Comment:
I think we should make this a `bool` on the v3 API and convert it when storing in v1 CRDs, since a bool is way more usable.

I think we should make this more clear what clients should do with this.

e.g.,

```
Deleted indicates whether or not this block affinity is disabled and is used as part of race-condition prevention. When set to true, clients should treat this block as if it does not exist.
```

Reply:
If we do that, then we won't be able to eventually change the API to use the same Spec objects (which makes it easier to do the conversion since all we need to do is copy the spec objects in). I know you mentioned wanting to go this route in follow-up PRs earlier, do we still want to follow that?